### PR TITLE
Leaves - Natalie and Sabrina

### PR DESCRIPTION
**Pull request by Galaxylaughing**
# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :------------- |
How did getting up to speed on the existing codebase go? If it went well, what worked? If it didn't go well, what will you do differently next time? | Overall it went well. It was initially very overwhelming because there are so many different things happening at once. When you look at one individual file, you cannot see the whole picture. Until you start to build methods that interact with other classes, you don't really understand how they are going to interact. |
| What inheritance relations exist between classes? | CSV_Record is the parent for Driver, Passenger, and Trip. What composition relations exist between classes? | One-to-many for Dispatcher to Driver and Passenger and Trip. One-to-many for Driver to Trip, and for Passenger to Trip. One-to-one for Driver to Passenger (per trip). One-one for Driver to Dispatcher. One-to-one for Passenger to Trip Dispatcher. Trip has a one-to-one relationship to Trip Dispatcher. Trip also has one-to-one relationship to Driver, and Passenger. |
Describe a decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | We had to make a decision about what to do if the cost of the ride was less than the fee per ride. We ended up deciding that negative "cost" (the driver reimburses the RideShare company) would be a possibility in our program. We thought about "you can't go below zero" meaning that if the cost field was less than the amount owed in fees to the RideShare company, that the cost field would automatically return 0. The third option we considered was, under the same circumstances, to not apply the fee payable to the RideShare company. Finally, we decided that the option that most closely reflects real-world scenarios was that a driver may owe the RideShare company for trips of extremely short duration (as a cost of doing business).
Give an example of a template method that you implemented for this assignment | We implemented the "from_csv method" that appears in the CSV_record class (as a private method) in driver.
Give an example of a nominal test that you wrote for this assignment | One nominal test was when we tested our request_trip method, testing for whether the start time is actually assigned as current time.
Give an example of an edge case test that you wrote for this assignment | One test that method request_trip should throw an error when no drivers are available.
What is a concept that you gained more clarity on as you worked on this assignment | Inheritance and also working with CSVs. Also working with some of the concepts introduced in POODR: the example we can think of is Trip Dispatcher class is a class that is primarily focused on sending messages. Generally, we have a lot more exposure to how classes can interact with one-another.

### CODE REVIEW

File: lib/trip.rb [hunk 1]
```diff
@@ -1,61 +1,96 @@
...
     end
-
+    
+    def connect_driver(driver)
+      @driver = driver
+      driver.add_trip(self)
+    end
+    
+    def duration
+      if end_time.nil? || start_time.nil?
```
Comment (beccaelenzil) [inline lib/trip.rb:69]: It seems that these `ArgumentError`s should be raised in the constructor when a trip object is instantiated since this is not an error with an argument passed to the duration method but instead and argument passed to the constructor.

File: lib/passenger.rb [hunk 2]
```diff
@@ -3,21 +3,41 @@
...
-
+    
     def add_trip(trip)
       @trips << trip
     end
-
+    
+    def net_expenditures
+      total_expenditure = 0
+      trips.each do |trip|
```
Comment (beccaelenzil) [inline lib/passenger.rb:21]: Consider using the .sum method.

File: lib/passenger.rb [hunk 3]
```diff
@@ -3,21 +3,41 @@
...
+        if trip.cost.nil?
+          total_expenditure += 0
+        else
+          total_expenditure += trip.cost
+        end
+      end
+      return total_expenditure
+    end 
+    
+    def total_time_spent
```
Comment (beccaelenzil) [inline lib/passenger.rb:31]: It looks like you're missing the logic to appropriately deal with incomplete trips.

File: test/passenger_test.rb [hunk 4]
```diff
@@ -1,74 +1,239 @@
...
+    
+    it "Is correctly caclulating the total time spent" do
+      expect(@passenger.total_time_spent).must_equal 3_300
+    end 
+    
+    it "Returns zero for no trips taken." do
+      gerald = RideShare::Passenger.new(id: 1, name: "Gerald", phone_number: "353-533-7678")
+      expect(gerald.total_time_spent).must_equal 0
+    end
+
```
Comment (beccaelenzil) [inline test/passenger_test.rb:201]: I think you need to also consider in process trips.

File: lib/trip_dispatcher.rb [hunk 5]
```diff
@@ -3,39 +3,96 @@
...
+    
+    def find_available_driver
+      all_available_drivers = []
+      drivers.each do |driver|
+        if driver.status == :AVAILABLE && driver.find_ongoing_trips.length == 0
+          all_available_drivers << driver   
+        end
+      end
+      if all_available_drivers.length == 0
+        raise ArgumentError, "No Available Drivers at this Time."
```
Comment (beccaelenzil) [inline lib/trip_dispatcher.rb:59]: This is not an ArgumentError because there are no bad arguments (`find_available_driver` doesn't take any arguments!). I would recommend using StandardError.

```suggestion
        raise StandardError, "No Available Drivers at this Time."
```

File: test/trip_dispatch_test.rb [hunk 6]
```diff
@@ -1,72 +1,74 @@
 require_relative 'test_helper'
```
Comment (beccaelenzil) [inline test/trip_dispatch_test.rb:1]: you dropped the "er" in the title of this file. It should be `trip_dispatcher_test.rb.` Sticking with this convention is important for the use of guard, and it will be very important in Rails to pay attention to naming coventions.

File: test/trip_dispatch_test.rb [hunk 7]
```diff
@@ -121,5 +122,151 @@ def build_test_dispatcher
         end
       end
     end
+    
+    describe "Find available Driver" do
```
Comment (beccaelenzil) [inline test/trip_dispatch_test.rb:126]: I had trouble finding the tests that checked on the functionality of .request_trip. You may consider calling attention to the fact you are testing request_trip in the `describe` string.

File: test/trip_dispatch_test.rb [hunk 8]
```diff
@@ -121,5 +122,151 @@ def build_test_dispatcher
...
+        }.must_raise ArgumentError
+      end
+    end
+    
+    describe "find find available driver intelligent assignment" do
+      before do
+        @dispatcher = build_test_dispatcher
+      end
+      
+      it "selects first the driver with no previous trips, then the driver with the least recent most-recent trip" do
```
Comment (beccaelenzil) [inline test/trip_dispatch_test.rb:168]: This is nicely organized with comments, but you may consider putting each of these expect in different it blocks.

**Comment by beccaelenzil** [comment]
# Ride Share
## What We're Looking For


Feature|Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | check
Answer comprehension questions | check -- it looks like you missed the question about composition and it is embedded in your answer about inheritance.
**Wave 1** |
Appropriate use of Ruby's `Time` | check
`Trip` has a helper method to calculate duration | Consider with `duration` is the correct place to check for valid input.
`Passenger` methods to calculate total cost and total time | check - see comment about incomplete trips
Complete tests for new behavior | check - see comment about incomplete trips
**Wave 2** |
`Driver` inherits from `CsvRecord`, and implements `from_csv` | check
`Driver`s and `Trip`s are connected by `TripDispatcher` | check
`Driver` has method to calculate average rating | check
`Driver` has a method to calculate total revenue | check
**Wave 3** |
`TripDispatcher` has a new method to create trips | check
Trip creation logic is correctly implemented | check
New `Trip` is connected to the appropriate `Driver` and `Passenger` | check 
Tests for `request_trip` |check
Wave 1 and 2 methods ignore in-progress trips | check -- take a look at total time
Tests for Wave 1 and 2 methods w/ in-progress trips | check -- take a look at total time
**Wave 4 (Optional)** |
`TripDispatcher` assigns trips to the correct driver |check
Complete tests |check
Appropriate helper methods were made to help with complex logic |check
**Overall** | Good job overall! Your code is well-organized and easy to read, and does a good job of solving the problem at hand. I've left a few inline comments for you to review -- many of them focus on naming conventions and test organization. It's clear that the learning goals for this assignment were met. Keep up the hard work!

File: lib/passenger.rb (hunk 3 above)
Comment (Galaxylaughing) [reply lib/passenger.rb:31]: I guess I'm confused about what we were supposed to do; we thought that by having the lines
```
if trip.cost.nil?
          total_expenditure += 0
```
that we were telling the program that if the trip cost was nil, which it would only be for an in-progress trip, then ignore it—add zero, thereby essentially doing nothing. That was our thought process anyway.
